core/src/main/java/org/elasticsearch/indices/memory/IndexingMemoryController.java

```diff
@@ -211,93 +179,181 @@ public ByteSizeValue translogBufferSize() {
-        try {
-            shard.updateBufferSize(shardIndexingBufferSize, shardTranslogBufferSize);
-        } catch (EngineClosedException | FlushNotAllowedEngineException e) {
-            // ignore
-        } catch (Exception e) {
-            logger.warn("failed to set shard {} index buffer to [{}]", e, shard.shardId(), shardIndexingBufferSize);
-        }
+    /** ask this shard to refresh, in the background, to free up heap */
+    protected void writeIndexingBufferAsync(IndexShard shard) {
+        shard.writeIndexingBufferAsync();
     }
 
-    /** check if any shards active status changed, now. */
+    /** used by tests to check if any shards active status changed, now. */
     public void forceCheck() {
         statusChecker.run();
     }
 
+    /** called by IndexShard to record that this many bytes were written to translog */
+    public void bytesWritten(int bytes) {
+        statusChecker.bytesWritten(bytes);
+    }
+
+    /** Asks this shard to throttle indexing to one thread */
+    protected void activateThrottling(IndexShard shard) {
+        shard.activateThrottling();
+    }
+
+    /** Asks this shard to stop throttling indexing to one thread */
+    protected void deactivateThrottling(IndexShard shard) {
+        shard.deactivateThrottling();
+    }
+
+    static final class ShardAndBytesUsed implements Comparable<ShardAndBytesUsed> {
+        final long bytesUsed;
+        final IndexShard shard;
+
+        public ShardAndBytesUsed(long bytesUsed, IndexShard shard) {
+            this.bytesUsed = bytesUsed;
+            this.shard = shard;
+        }
+
+        @Override
+        public int compareTo(ShardAndBytesUsed other) {
+            // Sort larger shards first:
+            return Long.compare(other.bytesUsed, bytesUsed);
+        }
+    }
+
     class ShardsIndicesStatusChecker implements Runnable {
+
+        long bytesWrittenSinceCheck;
+
+        /** Shard calls this on each indexing/delete op */
+        public synchronized void bytesWritten(int bytes) {
+            bytesWrittenSinceCheck += bytes;
+            if (bytesWrittenSinceCheck > indexingBuffer.bytes()/30) {
+                // NOTE: this is only an approximate check, because bytes written is to the translog, vs indexing memory buffer which is
+                // typically smaller but can be larger in extreme cases (many unique terms).  This logic is here only as a safety against
+                // thread starvation or too infrequent checking, to ensure we are still checking periodically, in proportion to bytes
+                // processed by indexing:
+                run();
+            }
+        }
+
         @Override
         public synchronized void run() {
-            List<IndexShard> availableShards = availableShards();
-            List<IndexShard> activeShards = new ArrayList<>();
-            for (IndexShard shard : availableShards) {
-                if (!checkIdle(shard)) {
-                    activeShards.add(shard);
+
```

Comment:
phew nice! I mean it's pretty much the same as IW or rather DocumentsWriter does on a per doc basis but it's a pretty large method with lots of loops I really wonder if we wanna only do it in a tryLock fashion if possible?